File: Hard/2026-04-12-1320-Minimum-Distance-to-Type-a-Word-Using-Two-Fingers/solution.js
var minimumDistance = function(word) {
    const positions = {};
    const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    
    for (let i = 0; i < alphabet.length; i++) {
        positions[alphabet[i]] = [Math.floor(i / 6), i % 6];
    }
    
    const n = word.length;
    const dp = Array.from({ length: n + 1 }, () => Array(27).fill(Infinity));
    dp[0][26] = 0; // Starting with both fingers free

    for (let i = 0; i < n; i++) {
        const [x1, y1] = positions[word[i]];
        for (let j = 0; j <= 26; j++) {
            if (dp[i][j] === Infinity) continue;
            // Move finger 1 to current letter
            const dist1 = j === 26 ? 0 : Math.abs(x1 - positions[alphabet[j]][0]) + Math.abs(y1 - positions[alphabet[j]][1]);
            dp[i + 1][i] = Math.min(dp[i + 1][i], dp[i][j] + dist1);
            // Move finger 2 to current letter
            const dist2 = j === 26 ? 0 : Math.abs(x1 - positions[alphabet[j]][0]) + Math.abs(y1 - positions[alphabet[j]][1]);
            dp[i + 1][j] = Math.min(dp[i + 1][j], dp[i][j] + dist2);
        }
    }

    return Math.min(...dp[n]);
};

File: Hard/2026-04-12-1320-Minimum-Distance-to-Type-a-Word-Using-Two-Fingers/solution.py
class Solution:
    def minimumDistance(self, word: str) -> int:
        # Keyboard layout coordinates
        keyboard = {chr(i): (i // 6, i % 6) for i in range(ord('A'), ord('Z') + 1)}
        
        n = len(word)
        dp = [[[float('inf')] * 26 for _ in range(26)] for _ in range(n + 1)]
        
        # Initial positions of fingers are free
        for i in range(26):
            dp[0][i][i] = 0
        
        for i in range(1, n + 1):
            curr_char = word[i - 1]
            curr_x, curr_y = keyboard[curr_char]
            for j in range(26):
                for k in range(26):
                    # Move finger 1 to current character
                    dp[i][ord(curr_char) - ord('A')][k] = min(dp[i][ord(curr_char) - ord('A')][k], 
                                                                dp[i - 1][j][k] + (abs(curr_x - keyboard[chr(j + ord('A'))][0]) + 
                                                                                   abs(curr_y - keyboard[chr(j + ord('A'))][1])))
                    # Move finger 2 to current character
                    dp[i][j][ord(curr_char) - ord('A')] = min(dp[i][j][ord(curr_char) - ord('A')], 
                                                                dp[i - 1][j][k] + (abs(curr_x - keyboard[chr(k + ord('A'))][0]) + 
                                                                                   abs(curr_y - keyboard[chr(k + ord('A'))][1])))
        
        return min(dp[n][j][k] for j in range(26) for k in range(26))